File: sudoku/src/sudoku.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>

#include <sudoku.h>
#include <sudodef.h>
#include <methods.h>
#include <sudomat.h>
#include <sudostep.h>

SUDOKU_API Sudoku*
CreateSudoku(SudokuReadData readData, SudokuWriteData writeData, SolveProcessCallback procCallback, void* data) {
  if (readData == NULL || writeData == NULL) {
    return NULL;
  }
  Sudoku* sudo = (Sudoku*)malloc(sizeof(Sudoku));
  if (sudo == NULL) {
    return NULL;
  }
  memset(sudo, 0, sizeof(Sudoku));
  sudo->readData = readData;
  sudo->writeData = writeData;
  sudo->procCallback = procCallback;
  sudo->dataIO = data;
  sudo->hasWriteData = false;
  sudo->unknownCount = 0;
  sudo->dancing = false;
  sudo->ansCallback = NULL;
  sudo->ansCount = 0;
  sudo->improveCount = 0;
  initial_all(sudo);
  return sudo;
}

SUDOKU_API void DestroySudoku(Sudoku* sudo) {
  if (sudo == NULL) {
    return;
  }
  free(sudo);
}

SUDOKU_API bool VerifySudoku(Sudoku* sudo) {
  int board[BOARD_SIZE];
  memset(board, 0, sizeof(int) * BOARD_SIZE);
  for (int i = 0; i < 9; i++) {
    for (int j = 0; j < 9; j++) {
      int index = i * 9 + j;
      board[index] = sudo->sudoData[i][j][0];
    }
  }
  return VerifySudokuBoard(board);
}

SUDOKU_API int GetKnownCount(Sudoku* sudo) {
  return BOARD_SIZE - sudo->unknownCount;
}

SUDOKU_API void MakeResultString(Sudoku* sudo, char* buffer, int size) {
  memset(buffer, 0, sizeof(char) * size);
  if (size < RESULT_BUFFER_SIZE) {
    return;
  }
  int writeCount = 0;
  for (int i = 0; i < 9; i++) {
    for (int j = 0; j < 9; j++) {
      sprintf(buffer + writeCount, "%d ", sudo->sudoData[i][j][0]);
      writeCount += 2;
      if (j == 2 || j == 5) {
        sprintf(buffer + writeCount, " ");
        writeCount += 1;
      }
    }
    sprintf(buffer + writeCount, "\n");
    writeCount += 1;
    if (i == 2 || i == 5) {
      sprintf(buffer + writeCount, "\n");
      writeCount += 1;
    }
  }
  assert(writeCount == RESULT_BUFFER_SIZE - 1);
}

SUDOKU_API int CalculateSudokuAll(Sudoku* sudo, bool dancing, SudokuAnswerCallback cb, void* data) {
  sudo->dancing = dancing;
  sudo->ansCallback = cb;
  sudo->data = data;
  sudo->ansCount = 0;
  if (sudo->dancing) {
    calculate_with_dancing(sudo);
  } else {
    calculate_step_by_step(sudo);
  }
  return sudo->ansCount;
}

SUDOKU_API bool VerifySudokuBoard(const int* board) {
  if (grids_has_zero(board)) {
    return false;
  }
  if (row_has_repeat_element(board)) {
    return false;
  }
  if (col_has_repeat_element(board)) {
    return false;
  }
  if (block_has_repeat_element(board)) {
    return false;
  }
  return true;
}

const char* solve_type_name[] = {"None",
                                 "GridOnlyOnePosibility",
                                 "RowNumberOnlyOneGrid",
                                 "ColNumberOnlyOneGrid",
                                 "BlockNumberOnlyOneGrid"};
static const char* improve_type_name[] = {"DoNotImprove",

                                          "Row2GridWith2Posibility",
                                          "Col2GridWith2Posibility",
                                          "Block2GridWith2Posibility",

                                          "Row3GridWith3Posibility",
                                          "Col3GridWith3Posibility",
                                          "Block3GridWith3Posibility",

                                          "Row2NumberIn2Grid",
                                          "Col2NumberIn2Grid",
                                          "Block2NumberIn2Grid",

                                          "Row3NumberIn3Grid",
                                          "Col3NumberIn3Grid",
                                          "Block3NumberIn3Grid",

                                          "InBlockNumberInOneRow",
                                          "InBlockNumberInOneCol",

                                          "InRowNumberInBlock",
                                          "InColNumberInBlock",

                                          "TwoRowOneNumberInTwoCol",
                                          "TwoColOneNumberInTwoRow",

                                          "ThreeRowOneNumberInThreeCol",
                                          "ThreeColOneNumberInThreeRow"};
SUDOKU_API const char* SolveTypeName(SolveType type) {
  return solve_type_name[(int)type];
}
SUDOKU_API const char* ImproveTypeName(ImproveType type) {
  return improve_type_name[(int)type];
}

File: demo/for-nodejs/sudoku.js
const refArray = require("ref-array");
const intArray = refArray("int");
const refStruct = require("ref-struct");

const libSudoku = require("./libsudoku");
const solveProcedure = refStruct({
  type: "int",
  gridOneX: "int", // [1, 9]
  gridOneY: "int",
  gridTwoX: "int",
  gridTwoY: "int",
  gridThreeX: "int",
  gridThreeY: "int",
  numberOne: "int", // [1, 9]
  numberTwo: "int",
  numberThree: "int",
  panelRow: "int", // [1, 3]
  panelCol: "int",
  number: "int", // [1, 9]
  line: "int", // [1, 9]
  rowOne: "int", // [1, 9]
  rowTwo: "int",
  rowThree: "int",
  colOne: "int",
  colTwo: "int",
  colThree: "int"
});
const ref = require("ref");
const solveProcedurePtr = ref.refType(solveProcedure);

module.exports = {
  CreateSudoku: function(readFunc, writeFunc, procFunc) {
    const ffi = require("ffi");
    // typedef int (*SudokuReadData)(void* data, int row, int col);
    const wrapReadFunc = ffi.Callback(
      "int",
      ["pointer", "int", "int"],
      function(data, row, col) {
        return readFunc(row, col);
      }
    );
    // typedef void (*SudokuWriteData)(void* data, int row, int col, int value, SolveType type);
    const wrapWriteFunc = ffi.Callback(
      "void",
      ["pointer", "int", "int", "int", "int"],
      function(data, row, col, value, type) {
        writeFunc(row, col, value, type);
      }
    );
    // typedef void (*SolveProcessCallback)(void* data, SolveProcedure* proc);
    const wrapProcFunc = ffi.Callback(
      "void",
      ["pointer", solveProcedurePtr],
      function(data, proc) {
        if (procFunc == null) {
          return;
        }
        const procStruct = new solveProcedure(proc);
        procFunc(procStruct);
      }
    );
    return {
      handler: libSudoku.CreateSudoku(
        wrapReadFunc,
        wrapWriteFunc,
        wrapProcFunc,
        wrapProcFunc
      ),
      DestroySudoku: function() {
        libSudoku.DestroySudoku(this.handler);
      },
      VerifySudoku: function() {
        return libSudoku.VerifySudoku(this.handler);
      },
      GetKnownCount: function() {
        return libSudoku.GetKnownCount(this.handler);
      },
      MakeResultString: function() {
        const RESULT_BUFFER_SIZE = 9 * 9 * 2 + 9 * 2 + 9 + 2 + 1;
        const buffer = new Buffer(RESULT_BUFFER_SIZE);
        libSudoku.MakeResultString(this.handler, buffer, RESULT_BUFFER_SIZE);
        return ref.readCString(buffer, 0);
      },
      CalculateSudokuAll: function(dancing, callback) {
        // typedef void (*SudokuAnswerCallback)(void* data, const char* ans);
        const wrapCallback = ffi.Callback(
          "void",
          ["pointer", "string"],
          function(data, ans) {
            if (callback == null) {
              return;
            }
            callback(ans);
          }
        );
        return libSudoku.CalculateSudokuAll(
          this.handler,
          dancing,
          wrapCallback,
          this.handler
        );
      }
    };
  },
  VerifySudokuBoard: function(board) {
    const boardArray = new intArray(board);
    return libSudoku.VerifySudokuBoard(boardArray.buffer);
  },
  SolveTypeName: function(type) {
    return libSudoku.SolveTypeName(type);
  },
  ImproveTypeName: function(type) {
    return libSudoku.ImproveTypeName(type);
  },
  SolveType: {
    None: 0,
    GridOnlyOnePosibility: 1,
    RowNumberOnlyOneGrid: 2,
    ColNumberOnlyOneGrid: 3,
    BlockNumberOnlyOneGrid: 4
  },
  ImproveType: {
    DoNotImprove: 0,

    Row2GridWith2Posibility: 1,
    Col2GridWith2Posibility: 2,
    Block2GridWith2Posibility: 3,

    Row3GridWith3Posibility: 4,
    Col3GridWith3Posibility: 5,
    Block3GridWith3Posibility: 6,

    Row2NumberIn2Grid: 7,
    Col2NumberIn2Grid: 8,
    Block2NumberIn2Grid: 9,

    Row3NumberIn3Grid: 10,
    Col3NumberIn3Grid: 11,
    Block3NumberIn3Grid: 12,

    InBlockNumberInOneRow: 13,
    InBlockNumberInOneCol: 14,

    InRowNumberInBlock: 15,
    InColNumberInBlock: 16,

    TwoRowOneNumberInTwoCol: 17,
    TwoColOneNumberInTwoRow: 18,

    ThreeRowOneNumberInThreeCol: 19,
    ThreeColOneNumberInThreeRow: 20
  }
};
